// logger/dist/logger.es2015.js
// Logger can be used and required from many places.
// This is global on / off switch for it, which all
// created logger contexts respect.

var onOff = {
  init() {
    window.filestackInternals.logger.working = false;
  },
  isWorking() {
    return window.filestackInternals.logger.working;
  },
  on() {
    window.filestackInternals.logger.working = true;
  },
  off() {
    window.filestackInternals.logger.working = false;
  },
};

/* eslint no-console:0 */

const context = (contextName, onOff) => {
  const api = function log(...stuff) {
    const convertedToStrings = [...stuff].map((thing) => {
      if (typeof thing === 'object') {
        return JSON.stringify(thing, (key, value) => {
          if (typeof value === 'function') {
            // If any function named json is found then call that function to get the json object.
            if (key === 'json') {
              try {
                return value();
              } catch (err) {
                // Throws? No worries. Just go on and return string.
              }
            }
            return '[Function]';
          }
          if (value instanceof File) {
            return `[File name: ${value.name}, mimetype: ${value.type}, size: ${value.size}]`;
          }
          return value;
        }, 2);
      }
      return thing;
    });
    if (onOff.isWorking()) {
      console.log(`[${contextName}]`, ...convertedToStrings);
    }
  };

  api.context = (subContextName) => {
    return context(`${contextName}][${subContextName}`, onOff);
  };

  api.on = onOff.on;
  api.off = onOff.off;

  return api;
};

const logger = context('filestack', onOff);

const initializeGlobalNamespace = () => {
  let namespace;
  if (typeof window === 'object') {
    namespace = window.filestackInternals;
    if (!namespace) {
      namespace = {};
      window.filestackInternals = namespace;
    }
    if (!namespace.logger) {
      namespace.logger = logger;
      onOff.init();
    }
  }
  return namespace;
};

initializeGlobalNamespace();

export default logger;
